source/extensions/clusters/logical_dns/logical_dns_cluster.cc

```diff
@@ -44,23 +47,58 @@ convertPriority(const envoy::config::endpoint::v3::ClusterLoadAssignment& load_a
 }
 } // namespace
 
-LogicalDnsCluster::LogicalDnsCluster(const envoy::config::cluster::v3::Cluster& cluster,
-                                     ClusterFactoryContext& context,
-                                     Network::DnsResolverSharedPtr dns_resolver,
-                                     absl::Status& creation_status)
+absl::StatusOr<std::unique_ptr<LogicalDnsCluster>>
+LogicalDnsCluster::create(const envoy::config::cluster::v3::Cluster& cluster,
+                          const envoy::extensions::clusters::dns::v3::DnsCluster& dns_cluster,
+                          ClusterFactoryContext& context,
+                          Network::DnsResolverSharedPtr dns_resolver) {
+  const auto& load_assignment = cluster.load_assignment();
+  const auto& locality_lb_endpoints = load_assignment.endpoints();
+  if (locality_lb_endpoints.size() != 1 || locality_lb_endpoints[0].lb_endpoints().size() != 1) {
+    if (cluster.has_load_assignment()) {
+      return absl::InvalidArgumentError(
+          "LOGICAL_DNS clusters must have a single locality_lb_endpoint and a single lb_endpoint");
+    } else {
+      return absl::InvalidArgumentError("LOGICAL_DNS clusters must have a single host");
+    }
+  }
+
+  const envoy::config::core::v3::SocketAddress& socket_address =
+      locality_lb_endpoints[0].lb_endpoints()[0].endpoint().address().socket_address();
+  if (!socket_address.resolver_name().empty()) {
+    return absl::InvalidArgumentError(
+        "LOGICAL_DNS clusters must NOT have a custom resolver name set");
+  }
+
+  absl::Status creation_status = absl::OkStatus();
+  std::unique_ptr<LogicalDnsCluster> ret;
+
+  ret = std::unique_ptr<LogicalDnsCluster>(new LogicalDnsCluster(
```

Comment:
@Stevenjin8 nit, maybe just std::make_unique and remove the indirection?

Reply:
the constructors are no public, but I can make them friend classes.

Reply:
I would have to make the make_unique function a friend, which seems like more effort than its worth